src/WinRT.Runtime2/InteropServices/RestrictedErrorInfo.cs

```diff
@@ -28,10 +34,151 @@ public static unsafe class RestrictedErrorInfo
     /// calls (both in managed and native code). This improves the debugging experience.
     /// </remarks>
     /// <seealso cref="Marshal.GetExceptionForHR(int)"/>
-    public static Exception? GetExceptionForHR(HRESULT errorCode)
+    public static Exception GetExceptionForHR(HRESULT errorCode, out bool restoredExceptionFromGlobalState)
     {
-        // TODO
-        return null;
+        restoredExceptionFromGlobalState = false;
+        Exception? exception;
+        string errorMessage = string.Empty;
+        Exception internalGetGlobalErrorStateException;
+
+        using WindowsRuntimeObjectReferenceValue restrictedErrorInfoValue = ExceptionHelpers.BorrowRestrictedErrorInfo();
+
+        try
+        {
+            void* restrictedErrorInfoValuePtr = restrictedErrorInfoValue.GetThisPtrUnsafe();
+
+            if (restrictedErrorInfoValuePtr != null)
+            {
+                if (WellKnownErrorCodes.Succeeded(IUnknownVftbl.QueryInterfaceUnsafe(
+                        restrictedErrorInfoValuePtr,
+                        in WellKnownInterfaceIds.IID_ILanguageExceptionErrorInfo,
+                        out void* languageErrorInfoPtr)))
+                {
+                    try
+                    {
+                        exception = ExceptionHelpers.GetLanguageException(languageErrorInfoPtr, errorCode);
+
+                        if (exception is not null)
+                        {
+                            restoredExceptionFromGlobalState = true;
+                            WindowsRuntimeObjectReference? restrictedErrorInfo = WindowsRuntimeObjectReference.Create(restrictedErrorInfoValuePtr, WellKnownInterfaceIds.IID_IRestrictedErrorInfo);
+
+                            if (restrictedErrorInfo != null)
+                            {
+                                ExceptionHelpers.AddExceptionDataForRestrictedErrorInfo(exception, restrictedErrorInfo, true);
+                            }
+
+                            return exception;
+                        }
+                    }
+                    finally
+                    {
+                        WindowsRuntimeObjectMarshaller.Free(languageErrorInfoPtr);
+                    }
+                }
+
+                WindowsRuntimeObjectReference? restrictedErrorInfoToSave = WindowsRuntimeObjectReference.Create(restrictedErrorInfoValuePtr, WellKnownInterfaceIds.IID_IRestrictedErrorInfo);
+
+                IRestrictedErrorInfoMethods.GetErrorDetails(restrictedErrorInfoValuePtr, out string? description, out HRESULT hrLocal, out string? restrictedError, out string? restrictedCapabilitySid);
+
+                string restrictedErrorReference = IRestrictedErrorInfoMethods.GetReference(restrictedErrorInfoValuePtr);
```

Comment:
This is unused. Did you mean to use it anywhere? If not, we should remove this and delete that method too.

Reply:
Oh wait in the original code, it's being used at the end like this

<img width="575" height="328" alt="image" src="https://github.com/user-attachments/assets/3bdb6fff-9d5d-4dcf-bd84-e8397071cf9b" />